### ISSUE TITLE
Plans to align with Futures ecosystem?

### ISSUE THREAD

**Issue by thedodd** [capnproto/capnproto-rust#114]
Given that the Rust async ecosystem now has Futures in the standard library and async/await is quickly approaching stable, it seems as though it would be quite beneficial to bring the Rust Cpn ecosystem inline with these efforts as well. 

Any plans to do this?

I would love to start using CPN instead of protobuf ... but this divergence away from where async Rust is going is a bit of a non starter. 

I would love to help out as well if the maintainers would be willing to provide some direction / plan of attack.

**Comment by thedodd** [comment]
@dwrensha thanks for the quick response. I would tend to agree. There are definitely many crates throughout the Rust ecosystem which are in a similar situation. Start with futures `0.1` and use the 1 -> 3 shim, or wait for `0.3` to stabilize?

What seems to be definitive is that futures is the way to go, probably with tokio as the underlying system to drive the futures themselves (among other things). 

A few ideas:
- We could start with `0.1` initially. Make whatever architectural changes needed, but keep a focus on the end goal of integration with the stabilized std futures and async/await patterns.
- We could start a new unstable release line of this lib which uses `0.3` (which I admittedly have not actually built anything with `0.3` yet). Then once everything stabilizes and no longer requires nightly, we can cut an official stable release.

Any way, mostly just tossing ideas around to see what you think. Let me know.

**Comment by dwrensha** [comment]
capnproto-rust already works with futures 0.1.  See the [rpc examples](https://github.com/capnproto/capnproto-rust/tree/f28b3014b824ce5aa0c31baf91a88ce035624397/capnp-rpc/examples), for instance. Is there any reason that the current level of support is inadequate for your needs, until futures 0.3 or std::future::Future stabilizes?

**Comment by josephg** [comment]
The futures 0.3 crate actually just [re-exposes core::future::Future](https://github.com/rust-lang-nursery/futures-rs/blob/a0f58c2767161986bbe094c07b1617cbff390458/futures-core/src/future/mod.rs#L6), adding executors and future combinators. The futures API it exposes is the API thats on track to become stable in std.

Thats not an argument for porting across to the new API immediately. But, if you get everything working against futures-preview 0.3 now in a branch, it should be simple enough to merge that change once the futures API is ready.

**Comment by thedodd** [comment]
A pretty interesting [blog post from boats](https://boats.gitlab.io/blog/post/romio/) on using async/await/futures0.3 with tokio right now. [The repo is here](https://github.com/withoutboats/romio).

Would be pretty cool to experiment with that for this system. If I get some time, I would love to do so. We'll see.

**Comment by Progdrasil** [comment]
Now that `std::future::Future` has been released into stable, is there any news when this library will refactor towards it?

**Comment by dwrensha** [comment]
I've been making steady progress on the [std-future](https://github.com/capnproto/capnproto-rust/tree/std-future) branch. As of now, almost all of the capnp-rpc tests pass, and the calculator and pubsub examples are working.

**Comment by AThilenius** [comment]
@dwrensha I'm still very new to the Rust futures scene (so with a grain of salt) but `async-std` has things like 'tasks' which manage an executor for you, so some of the example code cleans up a bit. Example:

```rust
fn main() -> Result<()> {
    let opt = Opt::from_args();

    task::block_on(accept_loop(opt.address))
}

async fn accept_loop(addr: impl ToSocketAddrs) -> Result<()> {
    let listener = TcpListener::bind(addr).await?;
    let mut incoming = listener.incoming();
    while let Some(stream) = incoming.next().await {
        let stream = stream?;
        task::spawn(async {
            serve_loop(stream).await.unwrap();
        });
    }
    Ok(())
}

async fn serve_loop(stream: TcpStream) -> Result<()> {
    let calc =
        calculator::ToClient::new(CalculatorImpl::default()).into_client::<::capnp_rpc::Server>();
    stream.set_nodelay(true)?;
    let (reader, writer) = stream.split();
    let network = twoparty::VatNetwork::new(
        reader,
        writer,
        rpc_twoparty_capnp::Side::Server,
        Default::default(),
    );

    let _rpc_system = Box::pin(RpcSystem::new(Box::new(network), Some(calc.clone().client)));

    Ok(())
}
```

**Comment by dwrensha** [comment]
I just merged the std-future branch into master: 0e825eecbf2337d1fb2caed015bfa4862a195d40

**Comment by dwrensha** [comment]
Released v0.11.0.

**Comment by josephg** [comment]
Awesome 💃
